core/src/data_sources/data_source.rs

```diff
@@ -594,37 +596,156 @@ impl DataSource {
+                                                        .data
+                                                        .iter()
+                                                        .map(|v| *v as f64)
+                                                        .collect::<Vec<f64>>(),
+                                                    model: self.config.model_id.clone(),
+                                                    provider: self.config.provider_id.to_string(),
+                                                },
+                                            );
+                                        }
+                                        _ => {}
+                                    },
+                                    None => {}
+                                }
+                            }
+                            _ => {}
+                        },
+                        None => {}
+                    }
+                });
+                page_offset = scroll_results.next_page_offset;
+                if page_offset.is_none() {
+                    break;
+                }
+            }
+            let cache_hits = splits
+                .iter()
+                .filter(|s| {
+                    let mut hasher = blake3::Hasher::new();
+                    hasher.update(document_hash.as_bytes());
+                    hasher.update(s.as_bytes());
+                    let hash = format!("{}", hasher.finalize().to_hex());
+                    cached_embeddings.contains_key(&hash)
+                })
+                .count();
+            utils::done(&format!(
+                "Embeddings cache hits: data_source_id={} document_id={} cache_hits={} splits={}",
+                self.data_source_id,
+                document_id,
+                cache_hits,
+                splits.len()
+            ));
+            if cache_hits != splits.len() {
+                should_embbed = true;
+            } else {
+                for s in splits.clone() {
+                    let mut hasher = blake3::Hasher::new();
+                    hasher.update(document_hash.as_bytes());
+                    hasher.update(s.as_bytes());
+                    let hash = format!("{}", hasher.finalize().to_hex());
+                    match cached_embeddings.contains_key(&hash) {
+                        true => {
+                            e.push((offset, s, cached_embeddings[&hash].clone()));
+                            offset += 1;
+                        }
+                        false => {
+                            should_embbed = true;
+                            e = vec![];
+                            break;
+                        }
+                    };
+                }
+            }
+        }
 
-            for (s, v) in chunk.into_iter().zip(v.into_iter()) {
-                e.push((offset, s, v));
-                offset += 1;
+        if should_embbed {
```

Comment:
So here we remove shoiuld_embed.

We take the splits, we remove the cached ones create a list of chunks we do need to re-embed and we chunk them and embed them and move them in result hash table

Finally you iterate on the chunks and pull them from the result and cache hash tables.